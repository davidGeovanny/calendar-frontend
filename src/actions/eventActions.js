import Swal from "sweetalert2";
import { types } from "../types/types";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";

export const eventStartAddNew = ( event ) => {
    return async ( dispatch, getState ) => {

        const { uid, name } = getState().auth;

        try {
            const resp = await fetchWithToken( 'events', event, 'POST' );
            const body = await resp.json();
        
            if( body.ok ) {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                };

                dispatch( eventAddNew( event ) );
            }  
        } catch (error) {
            console.log(error);
        }
    };
};

const eventAddNew = ( event ) => {
    return {
        type: types.eventAddNew,
        payload: event
    };
}

export const eventSetActive = ( event ) => {
    return {
        type: types.eventSetActive,
        payload: event
    };
}

export const eventClearActiveEvent = () => {
    return {
        type: types.eventClearActiveEvent,
    };
}

export const eventStartUpdate = ( event ) => {
    return async ( dispatch ) => {
        try {
            const resp = await fetchWithToken( `events/${ event.id }`, event, 'PUT' );
            const body = await resp.json();

            if( body.ok ) {
                dispatch( eventUpdated( event ) );
            } else {
                Swal.fire({
                    title: 'Error',
                    text: body.msg,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
        }
    };
};

const eventUpdated = ( event ) => {
    return {
        type: types.eventUpdated,
        payload: event
    };
}

export const eventStartDelete = () => {
    return async ( dispatch, getState ) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const resp = await fetchWithToken( `events/${ id }`, {}, 'DELETE' );
            const body = await resp.json();

            if( body.ok ) {
                dispatch( eventDeleted() );
            } else {
                Swal.fire({
                    title: 'Error',
                    text: body.msg,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
        }
    };
};

const eventDeleted = () => {
    return {
        type: types.eventDeleted,
    };
}

export const eventStartLoad = () => {
    return async ( dispatch ) => {
        try {
            const resp = await fetchWithToken( 'events' );
            const body = await resp.json();

            const events = prepareEvents( body.events );

            dispatch( eventLoaded( events ) );
        } catch (error) {
            console.log(error);
        }
    };
}

const eventLoaded = ( events ) => {
    return {
        type: types.eventLoaded,
        payload: events
    }
};

export const eventLogout = () => {
    return {
        type: types.eventLogout,
    }
};