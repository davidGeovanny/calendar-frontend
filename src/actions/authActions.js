import Swal from "sweetalert2";
import { types } from "../types/types";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { eventLogout } from "./eventActions";

export const startLogin = ( email, password ) => {
    return async ( dispatch ) => {
        const resp = await fetchWithoutToken( 'auth/', { email, password }, 'POST' );
        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login(
                { 
                    uid: body.uid, 
                    name: body.name 
                }
            ));
        } else {
            Swal.fire({
                title: 'Error',
                text: body.msg,
                icon: 'error' 
            });
        }
    };
};

const login = ( user ) => {
    return {
        type: types.authLogin,
        payload: user
    };
};

export const startRegister = ( name, email, password ) => {
    return async ( dispatch ) => {

        const resp = await fetchWithoutToken( 'auth/new', { name, email, password }, 'POST' );

        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login(
                { 
                    uid: body.uid, 
                    name: body.name 
                }
            ));
        } else {
            Swal.fire({
                title: 'Error',
                text: body.msg,
                icon: 'error',
            });
            dispatch( checkingFinish() );
        }
    };
};

export const startChecking = () => {
    return async ( dispatch ) => {
        const resp = await fetchWithToken( 'auth/renew' );

        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login(
                { 
                    uid: body.uid, 
                    name: body.name 
                }
            ));
        } else {
            dispatch( checkingFinish() );
        }
    };
};

const checkingFinish = () => {
    return {
        type: types.authCheckingFinish,
    };
};

export const startLogout = () => {
    return async ( dispatch ) => {
        localStorage.clear();
        dispatch( logout() );
        dispatch( eventLogout() );
    };
};

const logout = () => {
    return {
        type: types.authLogout,
    };
};