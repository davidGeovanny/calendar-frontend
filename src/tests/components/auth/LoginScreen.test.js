import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/authActions';

jest.mock( '../../../actions/authActions', () => {
    return {
        startLogin: jest.fn(),
        startRegister: jest.fn(),
    };
});

jest.mock( 'sweetalert2', () => {
    return {
        fire: jest.fn(),
    };
});

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initState = {};

const store = mockStore( initState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <LoginScreen />
    </Provider>
);

describe('Test with LoginScreen component', () => {

    beforeEach( () => {
        jest.clearAllMocks();
    });

    test('should match with the snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should call the dispatch(startLogin)', () => {
        const loginEmail = 'email@gmail.com';
        const loginPassword = '123456';

        wrapper.find('input[name="loginEmail"]').simulate('change', {
            target: {
                name: 'loginEmail',
                value: loginEmail
            }
        });
        wrapper.find('input[name="loginPassword"]').simulate('change', {
            target: {
                name: 'loginPassword',
                value: loginPassword
            }
        });

        wrapper.find('form').at(0).simulate('submit', {
            preventDefault() {},
        });

        expect( startLogin ).toHaveBeenCalledWith( loginEmail, loginPassword );
    });
    
    test('should not register if the passwords are differente', () => {
        wrapper.find('input[name="registerPassword"]').simulate('change', {
            target: {
                name: 'registerPassword',
                value: '123456',
            }
        });
        wrapper.find('input[name="registerPasswordRepeat"]').simulate('change', {
            target: {
                name: 'registerPasswordRepeat',
                value: '1234567',
            }
        });

        wrapper.find('form').at(1).simulate('submit', {
            preventDefault() {},
        });

        expect( startRegister ).not.toHaveBeenCalled();
        expect( Swal.fire ).toHaveBeenCalled();
    });

    test('should dispatch startRegister when the passwords are the same', () => {
        wrapper.find('input[name="registerPassword"]').simulate('change', {
            target: {
                name: 'registerPassword',
                value: '123456',
            }
        });
        wrapper.find('input[name="registerPasswordRepeat"]').simulate('change', {
            target: {
                name: 'registerPasswordRepeat',
                value: '123456',
            }
        });

        wrapper.find('form').at(1).simulate('submit', {
            preventDefault() {},
        });

        expect( Swal.fire ).not.toHaveBeenCalled();
        expect( startRegister ).toHaveBeenCalled();
    });
    
    
})
