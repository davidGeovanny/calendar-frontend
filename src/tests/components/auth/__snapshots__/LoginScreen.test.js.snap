// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test with LoginScreen component should match with the snapshot 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="David Geovanny"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="david.geovanny@hotmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPasswordRepeat"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
